### PR TITLE
Bfs optimize vertex-out-out query

### PR DESCRIPTION
**Pull request by javeme**

### CODE REVIEW

**Comment by javeme** [comment]
cassandra ci:
image

**Comment by javeme** [comment]
rocksdb ci:
image

**Comment by javeme** [comment]
cassandra:
```
[ERROR] Tests run: 684, Failures: 2, Errors: 0, Skipped: 21, Time elapsed: 481.763 s <<< FAILURE! - in com.baidu.hugegraph.core.CoreTestSuite
[ERROR] testQueryEdgesWithLimitOnMultiLevel(com.baidu.hugegraph.core.EdgeCoreTest)  Time elapsed: 0.359 s  <<< FAILURE!
java.lang.AssertionError: expected:<4> but was:<3>
	at com.baidu.hugegraph.core.EdgeCoreTest.testQueryEdgesWithLimitOnMultiLevel(EdgeCoreTest.java:1999)
[ERROR] testQueryEdgesAdjacentVerticesWithLimitAndFilterProp(com.baidu.hugegraph.core.EdgeCoreTest)  Time elapsed: 1.056 s  <<< FAILURE!
java.lang.AssertionError: expected:<9> but was:<11>
	at com.baidu.hugegraph.core.EdgeCoreTest.testQueryEdgesAdjacentVerticesWithLimitAndFilterProp(EdgeCoreTest.java:2398)
```
mysql
```java:
[ERROR] Tests run: 684, Failures: 2, Errors: 0, Skipped: 26, Time elapsed: 314.164 s <<< FAILURE! - in com.baidu.hugegraph.core.CoreTestSuite
[ERROR] testQueryEdgesWithLimitOnMultiLevel(com.baidu.hugegraph.core.EdgeCoreTest)  Time elapsed: 0.273 s  <<< FAILURE!
java.lang.AssertionError: expected:<4> but was:<3>
	at com.baidu.hugegraph.core.EdgeCoreTest.testQueryEdgesWithLimitOnMultiLevel(EdgeCoreTest.java:1999)
[ERROR] testQueryEdgesAdjacentVerticesWithLimitAndFilterProp(com.baidu.hugegraph.core.EdgeCoreTest)  Time elapsed: 0.312 s  <<< FAILURE!
java.lang.AssertionError: expected:<9> but was:<11>
	at com.baidu.hugegraph.core.EdgeCoreTest.testQueryEdgesAdjacentVerticesWithLimitAndFilterProp(EdgeCoreTest.java:2398)
```

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStepWithoutPath.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    public HugeVertexStepWithoutPath(final VertexStep<E> originalVertexStep) {
+        super(originalVertexStep);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected Traverser.Admin<E> processNextStart() {
+        /* Override super.processNextStart() */
+        if (this.parents == null) {
+            Function<Object, Traverser.Admin<Vertex>> nextFunc = null;
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStepWithoutPath.java:63]: use java.util.Function

Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStepWithoutPath.java:63]: remove `= null`, it's redudant

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStepWithoutPath.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            if (prev.getClass() == HugeVertexStepWithoutPath.class) {
+                // Just optimize, it's ok to call this.starts.next()
+                nextFunc = arg -> ((HugeVertexStepWithoutPath<Vertex>) prev)
+                                  .processNextStart();
+            } else {
+                nextFunc = arg -> this.starts.next();
+            }
+
+            // TODO: use BatchMapperIterator
+            List<Traverser.Admin<Vertex>> vertexes = new ArrayList<>();
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStepWithoutPath.java:74]: replace with vertices

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStepWithoutPath.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        Iterator<Edge> edges = this.edges(traversers);
+        return this.queryAdjacentVertices(edges);
+    }
+
+    private Iterator<Edge> edges(List<Traverser.Admin<Vertex>> traversers) {
+        assert traversers.size() > 0;
+
+        BatchConditionQuery batchQuery = new BatchConditionQuery(HugeType.EDGE);
+
+        for (Traverser.Admin<Vertex> t : traversers) {
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStepWithoutPath.java:145]: t -> traverser

**Review by Linary — CHANGES_REQUESTED** [review]

**Review by Linary — CHANGES_REQUESTED** [review]

**Review by zhoney — DISMISSED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStepByBatch.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        if (this.head == null && traversers.size() > 0) {
+            this.head = traversers.get(0);
+        }
+        boolean queryVertex = this.returnsVertex();
+        boolean queryEdge = this.returnsEdge();
+        assert queryVertex || queryEdge;
+        if (queryVertex) {
+            this.iterator = (Iterator<E>) this.vertices(traversers);
+        } else {
+            assert queryEdge;
```
[the remark this answers is not in the document]
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStepByBatch.java:108]: yes, it's just for comment

**Review by imbajin — APPROVED** [review]

**Review by zhoney — APPROVED** [review]

**Pull request merged**